File: src/solutions/java/N387FirstUniqueCharacter.java
package solutions.java;

import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static java.util.function.Function.identity;

public class N387FirstUniqueCharacter {

    public int firstUniqChar(String s) {
        Map<Character, Long> countByChar = s.chars()//
                .mapToObj(i -> (char) i)//
                .collect(Collectors.groupingBy(identity(), Collectors.counting()));
        return IntStream.range(0, s.length())//
                .filter(i -> countByChar.get(s.charAt(i)).equals(1L))//
                .findFirst()//
                .orElse(-1);
    }
}

File: src/solutions/scala/N387FirstUniqueCharacter.scala
package solutions.scala

case object N387FirstUniqueCharacter {

  def firstUniqChar(s: String): Int = {
    val countByChar = s.groupBy(identity).mapValues(_.length)
    s.zipWithIndex.find(p => countByChar(p._1) == 1).map(_._2).getOrElse(-1)
  }

  def firstUniqChar2(s: String): Int = {
    s.zipWithIndex
      .groupBy(_._1)
      .values
      .map(s => (s.head._2, s.size))
      .filter(p => p._2 == 1)
      .map(_._1)
      .reduceOption(_ min _)
      .getOrElse(-1)
  }

  def main(args: Array[String]): Unit = {
    assert(firstUniqChar("leetcode") == 0)
    assert(firstUniqChar("loveleetcode") == 2)
  }
}
